docs/references/pubsub-websocket-protobuf-subprotocol.md

```diff
@@ -0,0 +1,417 @@
+---
+layout: docs
+toc: true
+group: references
+---
+     
+## Protobuf PubSub WebSocket Subprotocol
+
+This document describes the subprotocol `protobuf.webpubsub.azure.v1`.
+
+When the client is using this subprotocol, both outgoing data frame and incoming data frame are expected to be **protobuf** payloads.
+
+- [Overview](#overview)
+- [Permissions](#permissions)
+- [Requests](#requests)
+    - [Join Group Request](#join)
+    - [Leave Group Request](#leave)
+    - [Publish to Group Request](#publish)
+    - [Send Events Request](#events)
+- [Responses](#responses)
+    - [Ack Response](#ack)
+    - [Message Response](#message)
+    - [System Response](#system)
+
+### Overview
+
+Subprotocol `protobuf.webpubsub.azure.v1` empowers the clients to do publish/subscribe directly instead of a round trip to the upstream server. We call the WebSocket connection with `protobuf.webpubsub.azure.v1` subprotocol a Protobuf PubSub WebSocket client.
+
+For example, in JS, a PubSub WebSocket client can be created using:
+```js
+// Protobuf PubSub WebSocket client
+var pubsub = new WebSocket('wss://test.webpubsub.azure.com/client/hubs/hub1', 'protobuf.webpubsub.azure.v1');
+```
+For a simple WebSocket client, the *server* is a MUST HAVE role to handle the events from clients. A simple WebSocket connection always triggers a `message` event when it sends messages, and always relies on the server-side to process messages and do other operations. With the help of the `protobuf.webpubsub.azure.v1` subprotocol, an authorized client can join a group using [join requests](#join) and publish messages to a group using [publish requests](#publish) directly. It can also route messages to different upstreams (event handlers) by customizing the *event* the message belongs using [event requests](#events).
+
+Currently, WebPubSub Service only support [proto3](https://developers.google.com/protocol-buffers/docs/proto3).
+
+### Permissions
+
+As you may have noticed when we describe the PubSub WebSocket clients, that a client can publish to other clients only when it is *authorized* to. The `role`s of the client determines the *initial* permissions the client have:
+
+| Role | Permission |
+|---|---|
+| Not specified | The client can send event requests.
+| `webpubsub.joinLeaveGroup` | The client can join/leave any group.
+| `webpubsub.sendToGroup` | The client can publish messages to any group.
+| `webpubsub.joinLeaveGroup.<group>` | The client can join/leave group `<group>`.
+| `webpubsub.sendToGroup.<group>` | The client can publish messages to group `<group>`.
+
+The server-side can also grant or revoke permissions of the client dynamically through REST APIs or server SDKs.
+
+### Requests
+
+All request messages follow the following protobuf format.
+
+```protobuf
```

Comment:
>protobuf [](http://example.com/codeflow?start=3&length=8)

move this to a seperate `proto` file? So that users can directly download the protocol file